src/inc/lib/e164.php

```diff
@@ -0,0 +1,1156 @@
+
+	/**
+	 * Parse a phone number to determine the country code and subscriber number.
+	 *
+	 * @param string $phone_number A phone number which should only contain numbers representing each digit (vanity phone numbers with letters cause issues).
+	 *
+	 * @return array|null An array with keys `country_code` and `subscriber_number` if a match is found, otherwise null.
+	 */
+	public static function parse_phone_number( string $phone_number ): ?array {
+		$country_code = static::get_country_code( $phone_number );
+		if ( empty( $country_code ) ) {
+			return null;
+		}
+		return static::parse_as_country_code( $country_code, $phone_number );
+	}
+
+	/**
+	 * Return a string by providing an array with keys `country_code` and `subscriber_number`.
+	 *
+	 * @param array $parsed_phone_number An array with keys `country_code` and `subscriber_number`.
+	 *
+	 * @return string The e.164 formatted phone number.
+	 */
+	public static function from_parsed_phone_number( array $parsed_phone_number ): string {
+		return implode(
+			array(
+				'+',
+				$parsed_phone_number['country_code'],
+				$parsed_phone_number['subscriber_number'],
+			)
+		);
+	}
+
+	/**
+	 * Get the country code from the phone number.
+	 *
+	 * @param string $phone_number A phone number which should only contain numbers representing each digit (vanity phone numbers with letters cause issues).
+	 *
+	 * @return string|null The country code of the phone number, or null if one could not be found.
+	 */
+	public static function get_country_code( string $phone_number ): ?string {
+		$offset = static::looks_like_e164_format( $phone_number ) ? 1 : 0;
+		for ( $length = 1; $length < 3; $length++ ) {
+			$possible_country_code = substr( $phone_number, $offset, $length );
+			$possible_match        = static::get_country_config( $possible_country_code );
+			if ( ! empty( $possible_match ) ) {
+				if ( static::verify_country_code_and_subscriber_length( $possible_country_code, $phone_number ) ) {
+					return $possible_country_code;
+				}
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Parse a phone number with a known country code.
+	 *
+	 * @param string $country_code The country code present in the phone number.
+	 * @param string $phone_number A phone number which should only contain numbers representing each digit (vanity phone numbers with letters cause issues).
+	 *
+	 * @return array|null An array with keys `country_code` and `subscriber_number` if a match is found, otherwise null.
+	 */
+	public static function parse_as_country_code( string $country_code, string $phone_number ): ?array {
+		$country_config = static::get_country_config( $country_code );
+		if ( empty( $country_config ) ) {
+			return false;
+		}
+
+		$phone_number                = static::post_process( $phone_number );
+		$offset                      = static::looks_like_e164_format( $phone_number ) ? 1 : 0;
+		$country_code_length         = strlen( $country_code );
+		$phone_number_length         = strlen( $phone_number );
```

Comment:
I really don't understand those few lines...
if it "doesn't look like e164_format", why are you still fetching the country code?

Reply:
That function just looks for the `+` at the beginning. In America, it's common to include the country code sometimes. "800" numbers are toll-free numbers, and are generally referred to as "1-800" numbers. Most phones require you to dial a `1` before dialing an area code like 800 but depending on your state, a 1 may not be required to phone "long distance" which could mean a number like `123-123-1234` (which is the same as `(1-123-123-1234).